### PR TITLE
Openssl bump

### PR DESCRIPTION
**Pull request by theuni**
Bump to 1.0.1k and cope with its more strict DER checks.

### CODE REVIEW

**Comment by gavinandresen** [comment]
ACK: code reviewed, and fixes unit test errors running on OSX with OpenSSL 1.0.1k

File: src/ecwrapper.cpp [hunk 1]
```diff
@@ -117,10 +117,20 @@ bool CECKey::SetPubKey(const unsigned char* pubkey, size_t size) {
 }
 
 bool CECKey::Verify(const uint256 &hash, const std::vector<unsigned char>& vchSig) {
-    // -1 = error, 0 = bad sig, 1 = good
-    if (ECDSA_verify(0, (unsigned char*)&hash, sizeof(hash), &vchSig[0], vchSig.size(), pkey) != 1)
+    // New versions of OpenSSL will reject non-canonical DER signatures. de/re-serialize first.
+    unsigned char *norm_der = NULL;
+    ECDSA_SIG *norm_sig = ECDSA_SIG_new();
+    const unsigned char* sigptr = &vchSig[0];
```
Comment (sipa) [inline src/ecwrapper.cpp]: use vchSig.begin(); don't use [0] unless it's guaranteed that the length is at least 1.

**Comment by sipa** [comment]
untested ACK

**Comment by jgarzik** [comment]
ut ACK

**Comment by theuni** [comment]
@sipa afaik the iterator isn't guaranteed to be a pointer to the element there. Added an early return instead, you ok with that?

**Comment by gmaxwell** [comment]
Tested ACK. (tests and a invalidate/reconsider loop back to 338221. Reindexed without checkpoints from 0 to 165k (still in progress))

**Comment by petertodd** [comment]
ut ACK

**Comment by fanquake** [comment]
Tested ACK. As mentioned by Gavin, verified this fixes the unit tests on OSX when building with 1.0.1k

**Comment by sawmarku** [comment]
Un

**Comment by pstratem** [comment]
tested ack, invalidated to 322000 reconsidered upto 327086 (still running)

**Comment by petertodd** [comment]
tested ACK, -checkpoints=0 reindexes on testnet and mainnet.

**Comment by john-connor** [comment]
This patch serves no purpose other than providing a "hack" around a flaw in the Bitcoin core. Previously OpenSSL did not look for invalid DER encoded data with possible trailing garbage but in the latest version it does. The actual problem is that Bitcoin doesn't generate DER encoded signatures and this is completely unrelated to OpenSSL. Now you are taking the non-DER data and DER encoding it before calling ECDSA_verify. Because of this the same operation will be performed twice for every signature verification. You MUST always DER encode your signature before passing it off to OpenSSL and this is nothing new. So this being "low" on OpenSSL's list is correct from my expert opinion as it is only performing validation and Bitcoin is passing invalid data.

ECDSA_verify:
int ECDSA_verify(int type, const unsigned char *dgst, int dgst_len,
		const unsigned char *sigbuf, int sig_len, EC_KEY *eckey)
 	{
	ECDSA_SIG *s;
	const unsigned char *p = sigbuf;
	unsigned char *der = NULL;
	int derlen = -1;
	int ret=-1;

	s = ECDSA_SIG_new();
	if (s == NULL) return(ret);
	if (d2i_ECDSA_SIG(&s, &p, sig_len) == NULL) goto err;
	/* Ensure signature uses DER and doesn't have trailing garbage */
	derlen = i2d_ECDSA_SIG(s, &der);
	if (derlen != sig_len || memcmp(sigbuf, der, derlen))
		goto err;
	ret=ECDSA_do_verify(dgst, dgst_len, s, eckey);
err:
	if (derlen > 0)
		{
		OPENSSL_cleanse(der, derlen);
		OPENSSL_free(der);
		}
	ECDSA_SIG_free(s);
	return(ret);
	}

**Comment by jgarzik** [comment]
@john-connor It is quite relevant in consensus systems, which must maintain bug-for-bug compatibility.

**Comment by john-connor** [comment]
@jgarzik The flaw is in Bitcoin, not OpenSSL.

**Comment by john-connor** [comment]
@jgarzik  I know the blockchain is flawed because of this problem but gavin and everybody else is blaming OpenSSL it seems. Accept the fact that Bitcoin core has a major flaw in that it does not DER encode signatures. Why not fix the problem and hard fork instead of hacking the code?

**Comment by gmaxwell** [comment]
@john-connor "Trailing garbage" is technically incorrect. What OpenSSL's ECDSA_verify previously supported was a very broad subset of BER.  For example, encoding R and S with leading zeros.  This is substantive, non-backwards compatible, arguably somewhat under-disclosed (seems that you were also unaware of the actual change, in that you refer to it as "trailing garbage"), breaking API change from the existing and widely known behavior.  Signatures come from the outside world (Bitcoin Core itself has never generated anything but strictly minimal encodings), and the decoding functions are OpenSSL internal.  As a result previously accepted encodings (and not just 'trailing garbage') will be rejected.  

While this may not matter for most applications, a switch to be more conservative is not something that can be done in an uncoordinated way for a consensus system. This is just the state of things, and not a question of "flaw". My description of the matter is just a statement of the facts about it.

**Comment by luke-jr** [comment]
The flaw was that Bitcoin relied on non-consensus-safe code (OpenSSL) for consensus. If OpenSSL does not want to guarantee consensus-compatible behaviour for invalid signatures, that's fine for them, but it breaks Bitcoin. The plan has been to close this issue in Bitcoin for a long time, and it would have been nice if OpenSSL took care breaking compatibility everywhere in a mere bugfix release. Note that a hardfork (or even a softfork) to fix this is not something we can do overnight, as we do not have any kind of authoritative control the protocol: making a change requires the consensus of every Bitcoin user, and coordination of that is a time consuming process that generally takes months.

Note that with Bitcoin (and I imagine most scenarios you'd be verifying a signature!), signatures are from an external source, so you can't say "just encode them right"... any signature-checking code really should assume the signature is untrusted data and possibly an attempt to perform a remote exploit.

Also note that Bitcoin Core has itself generated only DER-compatible signatures for a long time.

**Comment by john-connor** [comment]
@gmaxwell Right, Bitcoin has a flaw in that it does not and never did perform DER encoding of it's signatures and had it done so we would not be having this discussion.

**Comment by john-connor** [comment]
@luke-jr Wrong, OpenSSL is fine. Bitcoin is broke in this context. You can shift blame if you wish.

**Comment by gmaxwell** [comment]
> never did perform DER encoding 

I'm afraid your statements don't really make a lot of sense. Signatures emitted by Bitcoin core are correctly DER encoded.  OpenSSL provides a verify function which takes a string of bytes with an already encoded signature from the outside world. This is its documented behavior, and it's how OpenSSL itself uses it.  I would be surprised if you were able to find a _single_ prior user of that function which pre-parsed to verify or re-encoded the inputs before handing them to OpenSSL.

**Comment by john-connor** [comment]
@gmaxwell Wrong, Bitcoin has failed to ensure it DER encodes the signature properly since day one. OpenSSL did not break anything but instead brought this to light. Now as a hack you are DER encoding the data so OpenSSL won't reject it.

**Comment by jgarzik** [comment]
@john-connor That is like saying "HTTP failure to ensure..."   Bitcoin is a protocol.  Many different softwares generate signatures.  Bitcoin used OpenSSL to validate those signatures generated by $manywares.  That validation just changed without notice.

**Comment by john-connor** [comment]
@jgarzik The validation did not change. It just became more strict. The fact that non-DER data is being passed into OpenSSL is Bitcoin's bug. This hack-around is wasting CPU cycles and should be fixed properly after a certain block # way in the future with an organized hard fork. This will make sure that Bitcoin doesn't break again in the future for similar reasons.

**Comment by petertodd** [comment]
@john-connor More strict validation can definitely be a breaking change with security implications. The most obvious is DoS attacks, but in a system where a signature can be a right to something valuable it can result in monetary loss as well: http://www.mail-archive.com/bitcoin-development@lists.sourceforge.net/msg06686.html

In any case, in the future Bitcoin will likely move from OpenSSL to using libsecp256k1 library written by Pieter Wuille - a long-time Bitcoin Core developer - which is better suited to our unique needs.

> Instead bitcoin passing a raw sha256 hash that may or may not be DER encoded. What makes you assume that the hashes are in fact DER encoded? :)

You appear to be quite mistaken at what the code in question is actually doing; signatures != hashes.

**Comment by john-connor** [comment]
Where does Bitcoin core DER encode the signature properly?

**Comment by gmaxwell** [comment]
https://github.com/bitcoin/bitcoin/blob/master/src/secp256k1/src/ecdsa_impl.h#L90

**Comment by john-connor** [comment]
The problem with Bitcoin core is that secp256k1_ecdsa_sign_compact does NOT DER encode the output nor is it encoded into DER format at any point thereafter. This is why it became broken with the improved integrity checking that the latest OpenSSL brought. So "compact signatures" have always operated in a broken manner because of this flaw in the Bitcoin core. All crypto signatures that leave the wire or a stored to disk MUST always be encoded in some manner using DER or otherwise to maintain compatibility and platform neutrality. It should be fixed properly to ensure the security of the system IMHO.

**Comment by gmaxwell** [comment]
@john-connor No, thats incorrect. sign_compact signatures are not used inside the Bitcoin network, and are not verified (and _cannot_ be verified) by OpenSSL's ECDSA_verify.

Compact signatures encode different data, and are fundamentally incompatible with that encoding. They're also completely unrelated to this (as mentioned, not used on the Bitcoin network).

**Comment by john-connor** [comment]
@gmaxwell Nevertheless that is sloppy cryptographic work.

So please explain why signature verification is failing? What is wrong with them? You claim they are "properly DER encoded" so it that is true then this can't happen.

Something is being suppressed here...

**Comment by gmaxwell** [comment]
@john-connor There is nothing sloppy about it. I'm afraid you're operating well outside of your expertise here. I'd generally be happy to educate you (or anyone else), but your continued antagonism is tiring.  In this case; You're complaining about not using DER encoding for data which cannot be DER encoded and is never passed to a function that expects DER encoding.

> So please explain why signature verification is failing?

Because while Bitcoin core generates and has always generated strictly encoded valid DER third party implementations such as Armory, Blockchain.info, and MTGOX (among many others) previously generated signatures with excessive padding (e.g. additional leading zeros), from time to time old (or new) instances of this continue to crop up. In the pastthe establish software, based on OpenSSL, accepted since OpenSSL provided no method to distinguish these signatures from a non-excessively padded signature.

> can't happen

I'm afraid your concept of what can and cannot happen is too narrow for working in a public network, especially under adversarial conditions. Any sequence of bit that can be sent, probably will be sent, especially if they trigger interesting behavior.

**Comment by john-connor** [comment]
The Bitcoin core SHOULD remove this "padding" when it deserializes the data so the problem is immediately eliminated and does not make it into "the machinery". My assumption is that if you try to fix the problem then you'd have to hard fork the network. Now the Bitcoin blockchain remains polluted with invalid signatures that can never be verified without this "hack-around". Satoshi Nakamoto didn't understand networking too well nor the concepts of platform neutrality back then so I understand how this has happened, thanks.

**Comment by gmaxwell** [comment]
> The Bitcoin core SHOULD remove this "padding" when it deserializes the data

The deseralization is inside OpenSSL's ECDSA_verify, the padding in question is actually inside the DER signature after the length fields (its part of the numbers themselves). As your own comments above indicate, no one would expect to need to decode, then reencode (doing so would be arguably a bad practice), especially for data passed into a function when it's express purpose was to take data from the outside world. And, as I pointed out above you're not likely to find any user in the world who did this.

It certainly would have been preferable to not depend on the construction and behavior of OpenSSL's deseralizer, but without reading the OpenSSL code carefully you'd never know how much hidden behavior it has, none of this has to do with portability platform neutrality though; it's a question consensus consistency.

As far as fixes. We have more tools at our disposal than just hard forks.

**Comment by john-connor** [comment]
My final words on this:
http://vanillacoin.net/forums/viewtopic.php?pid=1967#p1967

**Comment by gmaxwell** [comment]
Your claims there continue to be confused and incorrect. Notice above that I pointed out "In the past the establish software"-- subsequently Bitcoin Core implemented default local node policy to not relay incorrectly encoded signatures (by testing the transactions with it's own strict DER decoder, https://github.com/bitcoin/bitcoin/blob/master/src/script/interpreter.cpp#L89) as the first step in a process to eliminate them.  This widely announced change delayed transactions for software which continued to produce them.

Do not mistake your ignorance about the technology or the history for any dishonesty on my part. Feel free to discontinue posting on our repository if you're unwilling to conduct yourself in a professional manner.

**Comment by john-connor** [comment]
@gmaxwell You seem to be a decent guy aside the slander you posted on bitcointalk.org but blaming the entire OpenSSL project because you failed to perform your own consistency checking on the signature is uncalled-for.

**Comment by rnicoll** [comment]
I may be misreading the OpenSSL documentation, however doesn't d2i_ECDSA_SIG always create a new ECDSA_SIG? So, given it's passed a reference to the pointer, the one allocated at https://github.com/bitcoin/bitcoin/pull/5634/files#diff-62cfc3439aea7461011d2394accbf82bR125 is in fact lost, if I'm reading this correctly?

**Comment by pstratem** [comment]
@rnicoll 

The documentation does say that.... but openssl itself seems to be calling ECDSA_SIG_new() before d2i_ECDSA_SIG.

Possibly there's a memory leak in openssl itself here.

**Comment by gmaxwell** [comment]
@rnicoll Objectively, the current version does not leak (I tested with valgrind.) OpenSSL itself allocates first, so preserving its own behavior seemed to be conservative. The code in question appears to check the nullness of the pointer and allocates only if null, meaning the documentation is actually dangerously incorrect (since if you believed the documentation you may send uninitialized memory in there and have it fail to allocate).

**Comment by rnicoll** [comment]
@gmaxwell Great, thanks for checking.

**Comment by pstratem** [comment]
d2i_ECDSA_SIG does indeed segfault if passed an uninitialized pointer.

**Comment by pstratem** [comment]
https://gist.github.com/pstratem/7d31cf5f35d78640247f

**Comment by abbradar** [comment]
Would there be a new point release or a patch for 0.9.3? This cannot be cleanly applied to 0.9.3 as-is (`ecwrapper.cpp` is missing), and I thought it would be better to ask anyway to avoid blindly patching critical piece of software.

**Comment by laanwj** [comment]
Try checking out and/or cherry-picking from the v0.9.4 tag, it has the same changes.

**Comment by abbradar** [comment]
I'll just checkout v0.9.4 then. Thanks!

**Pull request merged**